### DISCUSSION TITLE
What is the difference between SSG and static pages ? (pages router)

### DISCUSSION THREAD

**Discussion by arichard-info** [vercel/next.js#58593] [Help]
### Summary

I'm working on a e-commerce website and we have a mix of SSG pages (using getStaticProps), dynamic pages (using getServerSideProps) and static pages (pages not depending on data).

Lets take this build output as an exemple : 

```
Route (pages)                              Size     First Load JS
┌ ○ /                                      2.62 kB        80.6 kB
├   /_app                                  0 B              78 kB
├ ○ /404                                   182 B          78.2 kB
├ ○ /default                               252 B          78.3 kB
├ λ /dynamic                               275 B          78.3 kB
└ ● /static                                273 B          78.3 kB
+ First Load JS shared by all              78 kB
  ├ chunks/framework-66d32731bdd20e83.js   45.2 kB
  ├ chunks/main-fe8b314813e4df48.js        31.8 kB
  ├ chunks/pages/_app-8037b8dd0a48781d.js  298 B
  ├ chunks/webpack-8fa1640cc84ba8fe.js     750 B
  └ css/ef46db3751d8e999.css               20 B

ƒ Middleware                               26.7 kB

○  (Static)   prerendered as static content
●  (SSG)      prerendered as static HTML (uses getStaticProps)
λ  (Dynamic)  server-rendered on demand using Node.js
```

- `/default` page is not using data and is considered as "static"
- `/static` page is using `getStaticProps` and is considered as "SSG"
- `/dynamic` page is using `getServerSideProps` and is considered as `dynamic`

Why are static pages and SSG pages two different things ? 

The documentation says that by default, Next.js pre-renders pages that are not using data with static generation (https://nextjs.org/docs/pages/building-your-application/rendering/static-site-generation#static-generation-without-data)

I'm asking the question because I'm having a hard time trying to configure CDN cache for the application.
Here I have 3 different cache behavior : 
- `/default` don't have any Cache-Control header by default
- `/static` have a Cache-Control set by next.js and cannot be overriden : `s-maxage=31536000, stale-while-revalidate`
- `/dynamic` have a Cache-Control header and can be overriden : `private, no-cache, no-store, max-age=0, must-revalidate`

The default cache control set for static pages doesn't work for me. Because if the CDN follows this rule, the page will be cached for a year, even if the origin changes in the meantime.

I'd like to be able to have fine-grained cache-control management in my application, but this difference between static and SSG prevents me from doing so

**Comment by icyJoseph** [comment]
> SSG pages (using getStaticProps), dynamic pages (using getServerSideProps) and static pages (pages not depending on data).

I think, in this case, you are making a faulty distinction.

SSG has two forms, with and without data. 

So what you have is: Dynamic pages (using getServerSideProps) and static pages, where some might use `getStaticProps + getStaticPaths`, some use `getStaticProps` only, and use none. Never mind that `_app` usage of `getInitialProps` implies more complexity.

Then there's ISR, which is an enhancement on SSG, that allows Next.js to rebuild the page at runtime, like it did during build time.

What is, I think, typically tricky for SSG pages, is that to some extend, there's actually two versions of each page. Let's say we are talking about `/foo`, then we have:

- An HTML page, with all of its resources, scripts, and other accepts, fully complete - `foo.html` or `foo/index.html`
- A compound of JavaScript and JSON, where the JSON is the result of calling `getStaticProps`

The HTML page is used when a user lands on your application. That is, the entry point is `/foo`. The next server sends the `foo.html` (or depending on server configs and next.config.js `/foo/index.html`) file for that page, along with its static assets.

However, if the user, was already in your page, at another route, and now they click a link that takes them to the `/foo` page, then the HTML file is not what comes into play.

The default mechanism is that, as soon as the link pointing to `/foo` is within the viewport, Next.js makes a so called, prefetch request, for the JavaScript that drives `/foo`, and if `/foo` was made with `getStaticProps`, then a request for the JSON result is also started, so that we can render `<Foo {...fooPropsFromJSON} />`, roughly speaking.

So you have to control access to the `HTML` file, and the `JSON` files that are also made. Usually scripts are hashed by content, so you can probably cache those for a very long time. IIRC, the JSON is also hashed by build ID (?? I don't remember right now), but these come through the `_next/data` path on your server.
